flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/EvictingWindowOperatorContractTest.java

```diff
@@ -0,0 +1,2654 @@
+		when(mockTrigger.onProcessingTime(anyLong(), Matchers.<TimeWindow>any(), anyTriggerContext())).thenReturn(TriggerResult.FIRE_AND_PURGE);
+	}
+
+	/**
+	 * Verify that there is no late-date side output if the {@code WindowAssigner} does
+	 * not assign any windows.
+	 */
+	@Test
+	public void testNoLateSideOutputForSkippedWindows() throws Exception {
+
+		OutputTag<Integer> lateOutputTag = new OutputTag<Integer>("late"){};
+
+		WindowAssigner<Integer, TimeWindow> mockAssigner = mockTimeWindowAssigner();
+		Trigger<Integer, TimeWindow> mockTrigger = mockTrigger();
+		InternalWindowFunction<Iterable<Integer>, Void, Integer, TimeWindow> mockWindowFunction = mockWindowFunction();
+
+		OneInputStreamOperatorTestHarness<Integer, Void> testHarness =
+				createWindowOperator(mockAssigner, mockTrigger, 0L, intListDescriptor, mockWindowFunction, lateOutputTag);
+
+		testHarness.open();
+
+		when(mockAssigner.assignWindows(anyInt(), anyLong(), anyAssignerContext()))
+				.thenReturn(Collections.<TimeWindow>emptyList());
+
+		testHarness.processWatermark(0);
+		testHarness.processElement(new StreamRecord<>(0, 5L));
+
+		verify(mockAssigner, times(1)).assignWindows(eq(0), eq(5L), anyAssignerContext());
+
+		assertTrue(testHarness.getSideOutput(lateOutputTag) == null || testHarness.getSideOutput(lateOutputTag).isEmpty());
+	}
+
+	@Test
+	public void testLateSideOutput() throws Exception {
+
+		OutputTag<Integer> lateOutputTag = new OutputTag<Integer>("late"){};
+
+		WindowAssigner<Integer, TimeWindow> mockAssigner = mockTimeWindowAssigner();
+		Trigger<Integer, TimeWindow> mockTrigger = mockTrigger();
+		InternalWindowFunction<Iterable<Integer>, Void, Integer, TimeWindow> mockWindowFunction = mockWindowFunction();
+
+		OneInputStreamOperatorTestHarness<Integer, Void> testHarness =
+				createWindowOperator(mockAssigner, mockTrigger, 0L, intListDescriptor, mockWindowFunction, lateOutputTag);
+
+		testHarness.open();
+
+		when(mockAssigner.assignWindows(anyInt(), anyLong(), anyAssignerContext()))
+				.thenReturn(Collections.singletonList(new TimeWindow(0, 0)));
+
+		testHarness.processWatermark(20);
+		testHarness.processElement(new StreamRecord<>(0, 5L));
+
+		verify(mockAssigner, times(1)).assignWindows(eq(0), eq(5L), anyAssignerContext());
+
+		assertThat(testHarness.getSideOutput(lateOutputTag),
+				containsInAnyOrder(isStreamRecord(0, 5L)));
+
+		// we should also see side output if the WindowAssigner assigns no windows
+		when(mockAssigner.assignWindows(anyInt(), anyLong(), anyAssignerContext()))
+				.thenReturn(Collections.<TimeWindow>emptyList());
+
+		testHarness.processElement(new StreamRecord<>(0, 10L));
+
+		verify(mockAssigner, times(1)).assignWindows(eq(0), eq(5L), anyAssignerContext());
+		verify(mockAssigner, times(1)).assignWindows(eq(0), eq(10L), anyAssignerContext());
+
+		assertThat(testHarness.getSideOutput(lateOutputTag),
+				containsInAnyOrder(isStreamRecord(0, 5L), isStreamRecord(0, 10L)));
```

Comment:
Isn't the ordering of side outputs supposed to be deterministic? Why do we need to use `containsInAnyOrder`?